### PR TITLE
EHN: Convert run_routine from being run on a thread to a subprocess

### PR DESCRIPTION
**Pull request by YektaY**
This pull request is still a work in progress. 

The PR changes the core functionality of badger, such that the run_routine is being run on a subprocess instead of a thread. 

This is a draft PR because there are missing and broken tests as well as at least one known missing feature. 

Feature needed 
- [x] Checking of a critical constraint violations is also missing from this iteration of the run_routine_subprocess method. 

Tests still needed 
- [x] core_subprocess 

Tests that need to be fixed
- [x] test_home_page
- [x] test_routine 
- [x] test_gui_basic 
- [x] test_routine_editor 
- [x] test_routine_page
- [x] test_routine_runner
- [x] test_run_monitor

### CODE REVIEW

File: src/badger/core_subprocess.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        # External triggers
+        if stop_process.is_set():
+            raise BadgerRunTerminatedError
+        elif pause_process.is_set():
+            pause_process.wait()
+        else:
+            return 0  # continue to run
+'''
+
+def convert_to_solution(result: DataFrame, routine: Routine):
```
Comment (YektaY) [inline src/badger/core_subprocess.py:89]: add doc string to this method

File: src/badger/core_subprocess.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    solution = (vars, objs, cons, stas, is_optimal,
+                vocs.variable_names,
+                vocs.objective_names,
+                vocs.constraint_names,
+                vocs.observable_names)
+
+    return solution
+
+
+def run_routine_subprocess(queue, evaluate_queue, stop_process, pause_process, wait_event) -> None:
```
Comment (YektaY) [inline src/badger/core_subprocess.py:124]: need to update this doc string

File: src/badger/gui/default/components/routine_runner.py [hunk 3]
```diff
@@ -102,79 +123,79 @@ def run(self) -> None:
...
+            self.timer.stop()
+            self.signals.finished.emit()
+
+    def after_evaluate(self, results) -> None:
+        """
+        This method emits updates sent over from the subprocess running the routine on the evaluate_queue.  
+        """
+        if self.timer.isActive():
+            self.signals.progress.emit(results)
+            time.sleep(0.1)
```
Comment (roussel-ryan) [inline src/badger/gui/default/components/routine_runner.py:157]: double check if this sleep is needed, if it is, add a comment to describe why its needed

File: src/badger/gui/default/components/routine_runner.py [hunk 4]
```diff
@@ -102,79 +123,79 @@ def run(self) -> None:
...
+
+    '''
+    def save_init_vars(self) -> None:
```
Comment (roussel-ryan) [inline src/badger/gui/default/components/routine_runner.py:160]: remove commented code block

File: src/badger/gui/default/components/run_monitor.py [hunk 5]
```diff
@@ -547,15 +539,15 @@ def _configure_plot(self, plot_object, inspector, names):
...
-        )
+        self.routine_runner = routine_runner = BadgerRoutineSubprocess(self.process_manager, self.routine, False)
+
         routine_runner.signals.env_ready.connect(self.env_ready)
         routine_runner.signals.finished.connect(self.routine_finished)
         routine_runner.signals.progress.connect(self.update)
         routine_runner.signals.error.connect(self.on_error)
         routine_runner.signals.info.connect(self.on_info)
-
+        #routine_runner.signals.routine_data.connect(self.update_vocs)
```
Comment (roussel-ryan) [inline src/badger/gui/default/components/run_monitor.py:549]: remove comment

File: src/badger/gui/default/components/run_monitor.py [hunk 6]
```diff
@@ -613,10 +602,11 @@ def extract_timestamp(self, data=None):
 
         return data["timestamp"].to_numpy(copy=True)
 
-    def update(self):
+    def update(self, results):
```
Comment (roussel-ryan) [inline src/badger/gui/default/components/run_monitor.py:605]: type hint

Comment (roussel-ryan) [reply src/badger/gui/default/components/run_monitor.py:605]: docstring that describes what happens here at a high level

File: src/badger/gui/default/components/run_monitor.py [hunk 7]
```diff
@@ -630,12 +620,15 @@ def update(self):
...
-    def update_curves(self):
+    def update_curves(self, results=None):
         use_time_axis = self.plot_x_axis == 1
         normalize_inputs = self.x_plot_y_axis == 1
 
-        data_copy = self.routine.sorted_data
+        if results is not None:
+            self.routine.data = results
 
+        data_copy = self.routine.sorted_data
```
Comment (roussel-ryan) [inline src/badger/gui/default/components/run_monitor.py:630]: check to make sure sorted data is a copy

File: src/badger/gui/default/components/run_monitor.py [hunk 8]
```diff
@@ -748,14 +740,16 @@ def routine_finished(self):
...
-
+    
         try:
             # TODO: fill in the states
-            run = archive_run(self.routine, states=None)
-            self.routine_runner.run_filename = run['filename']
-            env = self.routine.environment
-            path = run['path']
-            filename = run['filename'][:-4] + 'pickle'
+            if not self.testing:
```
Comment (roussel-ryan) [inline src/badger/gui/default/components/run_monitor.py:746]: add TODO here to use something other than a testing attribute

File: src/badger/gui/default/components/run_monitor.py (hunk 7 above)
Comment (YektaY) [reply src/badger/gui/default/components/run_monitor.py:630]: sorted data is a copy 

`  @property
    def sorted_data(self):
        data_copy = deepcopy(self.data)
        if data_copy is not None:
            data_copy.index = data_copy.index.astype(int)
            data_copy.sort_index(inplace=True)

        return data_copy`

File: src/badger/gui/default/components/routine_runner.py [hunk 9]
```diff
@@ -1,44 +1,42 @@
...
 from ....tests.utils import get_current_vars
-
+import torch 
 
 class BadgerRoutineSignals(QObject):
     env_ready = pyqtSignal(list)
     finished = pyqtSignal()
-    progress = pyqtSignal()
+    progress = pyqtSignal(object)
+    routine_data = pyqtSignal(VOCS)
```
[the remark this answers is not in the document]
Comment (YektaY) [reply src/badger/gui/default/components/routine_runner.py:20]: looks like this signal isn't being used anymore, I have removed it

File: src/badger/gui/default/components/routine_runner.py (hunk 3 above)
Comment (YektaY) [reply src/badger/gui/default/components/routine_runner.py:157]: it was no longer needed, was removed

File: src/badger/gui/default/components/routine_runner.py [hunk 10]
```diff
@@ -102,79 +123,79 @@ def run(self) -> None:
...
+        self.timer.start() 
+        
+    def check_queue(self) -> None:
+        """
+        This method checks the subprocess for updates in the evaluate_queue. 
+        It is called by a QTimer every 100 miliseconds.  
+        """
+        if not self.evaluate_queue.empty():
+            while not self.evaluate_queue.empty():                                                  
+                results = self.evaluate_queue.get()
```
[the remark this answers is not in the document]
Comment (YektaY) [reply src/badger/gui/default/components/routine_runner.py]: Switched to a Pipe, which resolved the crashing

File: src/badger/core_subprocess.py [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+                    max_eval = tc_config['max_eval']
+                    if len(routine.data) >= max_eval:
+                        raise BadgerRunTerminatedError
+                elif idx == 1:
+                    max_time = tc_config['max_time']
+                    dt = time.time() - start_time 
+                    if dt >= max_time:
+                        raise BadgerRunTerminatedError
+
+            # External triggers
```
[the remark this answers is not in the document]
Comment (YektaY) [reply src/badger/core_subprocess.py:191]: Left current method in place, might revisit this in a future PR

**Review by roussel-ryan — APPROVED** [review]

**Pull request merged**